src/endpoints/groups.js

```diff
@@ -0,0 +1,100 @@
+const authz = require("../authz");
+const {Group} = require("../groups");
+
+const {contentTypesProvided, contentTypesConsumed} = require("../negotiate");
+const {deleteByUrls, proxyFromUpstream, proxyToUpstream} = require("../upstreamRequestHandlers");
+
+
+const setGroup = (nameExtractor) => (req, res, next) => {
+  const group = new Group(nameExtractor(req));
+
+  authz.assertAuthorized(req.user, authz.actions.Read, group);
+
+  req.context.group = group;
+  return next();
+};
+
+
+/* Group customizations
+ */
+
+
+/* Group overview
+ */
+
+
+/* GET
+ */
+const getGroupOverview = contentTypesProvided([
+  ["text/markdown", sendGroupOverview],
+  ["text/plain", sendGroupOverview],
+]);
+
+
+/* PUT
+ */
+const putGroupOverview = contentTypesConsumed([
+  ["text/markdown", receiveGroupOverview],
+]);
+
+
+/* DELETE
+ */
+const deleteGroupOverview = async (req, res) => {
+  authz.assertAuthorized(req.user, authz.actions.Write, req.context.group);
+
+  // TODO: somehow have a req.context.group.urlFor?
+  const groupSource = req.context.source;
```

Comment:
Conceptually, the Group overview should be retrieved from the `Group` rather than the `GroupSource`. However, with the current functionality, there isn't a way to retrieve the S3 object URL from the `Group` itself. It just happens to work with `GroupSource.urlFor`, even though I assume the intended use case is for datasets/narratives. `Group.urlFor` is achievable though, see 3908f012d5136d3e01ba211c7f8f681257948c9a as an example.

Or, for a bigger refactor, should the method of storage (S3 here) be its own entity that's shared between both the `Group` and `GroupSource`?

Reply:
Consider this, for example: https://gist.github.com/tsibley/a42514eca2a44a73e5bf7eb33cfe3235

Reply:
Ok, this has been addressed by 82e3909318e6244a9c13dddc335721fdb198e721.

---

Re: your patch, I applied parts of it in 39d4db2fb6a5e7be49edd239427609a4d4296329...055efa733fdf52ea237b11f43198345afb242831. However, I'm a bit unsure about adding the Group `authz.actions.Read` assertion on `/groups/:groupName/settings`. Would that also suggest removing the existing read assertions on specific endpoints such as this?

https://github.com/nextstrain/nextstrain.org/blob/1de0bc91faf8eb21c4be7f2234b20b085950649a/src/endpoints/groups.js#L92-L98

I prefer keeping the calls to `assertAuthorized` closer to the invocation like how it is currently.

I also prefer having the calls to `assertAuthorized` closer to each other like how both `sendGroupOverview` and `receiveGroupOverview` have calls. I'd be more OK with the suggestion if read was the only option under `/groups/:groupName/settings`.

Reply:
> Would that also suggest removing the existing read assertions on specific endpoints such as this?

No. It wasn't my intention to suggest that.

> I prefer keeping the calls to `assertAuthorized` closer to the invocation like how it is currently. … I also prefer having the calls to `assertAuthorized` closer to each other…

Yes, agreed.

Reply:
This should be resolved with c3275458fc8bddc667dabe4163820b7ca56e280a.